lib/web_ui/lib/src/engine/semantics/semantics.dart

```diff
@@ -397,30 +412,185 @@ enum Role {
+  void addLabelAndValue() {
+    if (semanticsObject.hasLabel || semanticsObject.hasValue || semanticsObject.hasTooltip) {
+      addSecondaryRole(LabelAndValue(semanticsObject));
+    }
+  }
+
+  /// Adds generic functionality for handling taps and clicks.
+  void addTappable() {
+    if (semanticsObject.isTappable) {
+      addSecondaryRole(Tappable(semanticsObject));
+    }
+  }
+
+  /// Adds a secondary role to this primary role manager.
+  ///
+  /// This method should be called by concrete implementations of
+  /// [PrimaryRoleManager] during initialization.
+  @protected
+  void addSecondaryRole(RoleManager secondaryRoleManager) {
+    assert(
+      _secondaryRoleManagers?.any((RoleManager manager) => manager.role == secondaryRoleManager.role) != true,
+      'Cannot add secondary role ${secondaryRoleManager.role}. This object already has this secondary role.',
+    );
+    _secondaryRoleManagers ??= <RoleManager>[];
+    _secondaryRoleManagers!.add(secondaryRoleManager);
+  }
+
+  /// Called immediately after the fields of the [semanticsObject] are updated
+  /// by a [SemanticsUpdate].
+  ///
+  /// A concrete implementation of this method would typically use some of the
+  /// "is*Dirty" getters to find out exactly what's changed and apply the
+  /// minimum DOM updates.
+  ///
+  /// The base implementation requests every secondary role manager to update
+  /// the object.
+  @mustCallSuper
+  void update() {
+    final List<RoleManager>? secondaryRoles = _secondaryRoleManagers;
+    if (secondaryRoles == null) {
+      return;
+    }
+    for (final RoleManager secondaryRole in secondaryRoles) {
+      secondaryRole.update();
+    }
+  }
+
+  /// Whether this role manager was disposed of.
+  bool get isDisposed => _isDisposed;
+  bool _isDisposed = false;
+
+  /// Called when [semanticsObject] is removed, or when it changes its role such
+  /// that this role is no longer relevant.
+  ///
+  /// This method is expected to remove role-specific functionality from the
+  /// DOM. In particular, this method is the appropriate place to call
+  /// [EngineSemanticsOwner.removeGestureModeListener] if this role reponds to
+  /// gesture mode changes.
+  @mustCallSuper
+  void dispose() {
+    semanticsObject.clearAriaRole();
+    _isDisposed = true;
+  }
+}
+
+/// A role used when a more specific role couldn't be assigned to the node.
+final class GenericRole extends PrimaryRoleManager {
+  GenericRole(SemanticsObject semanticsObject) : super.withBasics(PrimaryRole.generic, semanticsObject);
+
+  @override
+  void update() {
+    super.update();
+
+    if (!semanticsObject.hasLabel) {
+      // The node didn't get a more specific role, and it has no label. It is
+      // likely that this node is simply there for positioning its children and
+      // has no other role for the screen reader to be aware of. In this case,
+      // the element does not need a `role` attribute at all.
+      return;
+    }
```

Comment:
Yep. That's actually a very common case. We have container nodes whose job is to position their children (e.g. apply offsets and other transforms), but otherwise they have no semantic role to play. In that case they don't get any `role` attribute at all.